plutus-core/testlib/PlutusCore/Generators/PIR/GenerateTypes.hs

```diff
@@ -0,0 +1,501 @@
+{-# LANGUAGE BangPatterns          #-}
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DataKinds             #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE DeriveFunctor         #-}
+{-# LANGUAGE DerivingVia           #-}
+{-# LANGUAGE FlexibleContexts      #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE GADTs                 #-}
+{-# LANGUAGE ImportQualifiedPost   #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NumericUnderscores    #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE PartialTypeSignatures #-}
+{-# LANGUAGE PatternSynonyms       #-}
+{-# LANGUAGE RecordWildCards       #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TupleSections         #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE UndecidableInstances  #-}
+{-# LANGUAGE ViewPatterns          #-}
+{-# OPTIONS_GHC -Wno-name-shadowing #-}
+{-# OPTIONS_GHC -Wno-incomplete-uni-patterns #-}
+{-# OPTIONS_GHC -Wno-partial-type-signatures #-}
+{-# OPTIONS_GHC -Wno-orphans #-}
+{-# OPTIONS_GHC -Wno-redundant-constraints #-}
+
+module PlutusCore.Generators.PIR.GenerateTypes where
+
+import Control.Monad.Reader
+
+import Data.Map (Map)
+import Data.Map qualified as Map
+import Data.Set (Set)
+import Data.Set qualified as Set
+import Data.String
+import GHC.Stack
+import Test.QuickCheck
+
+import PlutusCore.Builtin
+import PlutusCore.Default
+import PlutusCore.Generators.PIR.Common
+import PlutusCore.Name
+import PlutusCore.Normalize
+import PlutusCore.Quote (runQuoteT)
+import PlutusIR
+import PlutusIR.Error
+import PlutusIR.Subst
+
+import PlutusCore.Generators.PIR.GenTm
+
+{- Note [Debugging generators that don't generate well-typed/kinded terms/types]
+    This module implements generators for well-typed terms and well-kinded types.
+    If you came here after a property like `prop_genTypeCorrect` failed and you
+    didn't get a minimal counterexample (because shrinking requries well-typed terms)
+    you need to use a different trick. One trick that often works is to add the well-typedness
+    check in the generators - e.g. in `genTerm` you can do something like this:
+    ```
+    genTerm ... = myCheck $ do
+      ...
+      where
+        myCheck gen = do
+          (trm, type) <- gen
+          if notConformingToExpectedInvariant trm type then
+            error (show trm ++ " " ++ show type)
+          else
+            return (trm, type)
+    ```
+-}
+
+-- * Dealing with fresh names
+
+-- | Get the free variables in a type that appear in negative position
+negativeVars :: Type TyName DefaultUni () -> Set TyName
+negativeVars ty = case ty of
+  TyFun _ a b      -> positiveVars a <> negativeVars b
+  TyApp _ a b      -> negativeVars a <> negativeVars b
```

Comment:
You're probably not wrong on the correctness issue.

As it stands the functions are only used to ensure that we don't generate negative datatypes - as these are not particularly useful for real applications and mostly cause our static analysis (that we wrote this generator for) to loop forever (it loops forever because the program would loop forever because the program is encoding omega using negative datatypes). It might be worth-while to add an option allowing you to generate negative datatypes if you really wanted to.

Reply:
> As it stands the functions are only used to ensure that we don't generate negative datatypes

But if we agree it has this correctness issue, how can we be sure it prevents the generator from producing negative datatypes? For example can it potentially produce this (Agda-like pseudocode):

```
data Fine a where
  Fine : (a -> integer) -> Fine a

data NotFine where
  NotFine : Fine NotFine -> NotFine
```

? `negativeVars` won't recognize `NotFine` as being negative due to it having a "positive-looking" polarity.

> mostly cause our static analysis (that we wrote this generator for) to loop forever (it loops forever because the program would loop forever because the program is encoding omega using negative datatypes

But we do have recursion in our real-world programs, hence we do want to have it in tests. Does your generator not generate recursive programs? Or only some subset of them?

> mostly cause our static analysis (that we wrote this generator for) to loop forever 

Can't you just add an iteration limit or something? Recognize recursion and not touch it? Fold what was unfolded with tricks that people use for supercompilation?

Reply:
Ok, looking back on this I think the right solution is to drop the negativity check and possibly re-implement it properly later.

Reply:
Great! Thanks a lot.